### DISCUSSION TITLE
How to create custom OpenThread message on OpenThread NCP.

### DISCUSSION THREAD

**Discussion by Vanperdung** [openthread/openthread#10757] [Q&A] [answered]
Hello OpenThread Team,
I'm currently working with OpenThread NCP firmware on EFR32. To meet my specific requirement, I need to create a custom Spinel command and also OpenThread message. However, I couldn't find any examples or references related to creating custom messages in OpenThread NCP. 
Could you please provide guidance or point me to any resources on how to:

- Create a custom OpenThread message on NCP.
- Modify the NCP to support sending and receiving this custom message.

Thanks.

**Comment by abtink** [comment]
You can check the `OPENTHREAD_ENABLE_NCP_VENDOR_HOOK` feature and the related methods and their documentation:
https://github.com/openthread/openthread/blob/b42be4cff14595a66f603fc83460505c176754f7/src/ncp/ncp_base.hpp#L554-L614

You can also check [src/ncp/example_vendor_hook.cpp](https://github.com/openthread/openthread/blob/b42be4cff14595a66f603fc83460505c176754f7/src/ncp/example_vendor_hook.cpp) which is an example of how vendor hook can be implemented.

There are allocated ranges for `SPINEL_CMD`, `SPINEL_PROP`, `SPINEL_STATUS` for vendor specific use. You can check `spinel.h` header.

**Reply by Vanperdung** [reply]
Hi @abtink,
Thank you for your suggestion! I successfully added a new Spinel command. However, I’m now unsure how to send and receive OpenThread messages between two NCPs in the handler. Could you please provide some suggestions?

**Reply by abtink** [reply, marked as answer]
> how to send and receive OpenThread messages between two NCPs in the handler.

Spinel is a protocol for communication between two components on the same device, such as an NCP and a host through an interface like UART or SPI. It's not designed for communication between two NCPs (i.e., between two devices). 

It is not clear to me what your question is.

- If you want to learn how to implement a vendor-specific NCP/spinel commands (send commands, etc), the best place to look and learn from would be the `ncp_base_*` source files and how different command/get/set handlers are implemented.

- If you want to communicate between two devices, you can use IPv6 messages. Thread is a networking protocol that provides IPv6 support. Check the `openthread/ip6.h` or the `openthread/udp6.h` header for IPv6/UDP OpenThread APIs.
